ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/caching/CachingDamlLedgerStateReaderWithFingerprints.scala

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.validator.caching
+
+import com.daml.caching.Cache
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{DamlStateKey, DamlStateValue}
+import com.daml.ledger.participant.state.kvutils.{DamlKvutils, Fingerprint}
+import com.daml.ledger.validator.StateKeySerializationStrategy
+import com.daml.ledger.validator.preexecution.DamlLedgerStateReaderWithFingerprints
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class CachingDamlLedgerStateReaderWithFingerprints(
+    val cache: Cache[DamlStateKey, (DamlStateValue, Fingerprint)],
```

Comment:
We only want to store values that are available in the cache (hence we don't store an option). A simplification that I can see is switching to `Option[(DamlStateValue, Fingerprint)]` from `(Option[DamlStateValue], Fingerprint)` as we'll have the same fingerprint for empty values anyway hence wouldn't need to store such fingerprints unnecessarily (and this would also simplify the code below).
What do you think?

Reply:
I'd be wary about making assumptions about the fingerprints of reading `None`. Underlying implementations might use the same fingerprint or they might not. As an example think about a key-value store that uses the last-modified-time of a value was as the fingerprint (assuming every transaction is committed at a different time). In that case, a `None` value might be due to the key never having been written, or it might be due to the key having been deleted in some transaction.

Mentally, the model I'm using is that we are caching results of lookups from the backing store; and these lookups can return `Option[DamlStateValue]`. This allows the non-existence of a key to mean something as well.

Reply:
Makes sense, however, this design has one con: we will cache failed lookups for immutable objects such as party allocations and package uploads. This way the committer node will have all submissions fail that depend on a yet to be uploaded party or package. I.e., if the party allocation required for a submission is being processed after the submission itself on a committer node then the committer won't be able to process the submission till the cache entry expires for the failed party lookup. Do you think we should handle this corner case?

Reply:
OK. I see. What we are saying is that only positive lookups for immutable keys are to be cached. That makes sense, as they are immutable in an append-only fashion. Not immutable by default.

Thanks for raising this. If you could add a comment as to why we only cache positive lookups that would be helpful

Reply:
Done.